<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN"
   "http://www.w3.org/TR/html4/strict.dtd">
<HTML>
   <HEAD>
   	  <link rel="stylesheet" type="text/css" href="docs.css">
      <TITLE>SimGenome: simulated directed evolution of gene regulatory circuits</TITLE>
   </HEAD>
<BODY>


<h1>SimGenome</h1>
<p>Simulated directed evolution of gene regulatory circuits</p>
<p>Written by Victor Hanson-Smith :: EMAIL victor.hanson-smith@ucsf.edu</p>

<hr>`
<p><a href="manual.html">Overview</a> | <a href="installation.html">Download and Install</a> | <a href="tutorial.html">Tutorial</a> | <a href="cli.html">Command Reference</a></p>
<hr>

<h2>Required Parameters:</h2>

<p><strong>--outdir [path]</strong>, where [path] is a directory into which all output from this
simulation will be written.  By default, SimGenome will overwrite any previous contents in this folder, and
therefore it is strongly suggested that this folder not be shared with other
data or previous analaysis.  If you would like SimGenome to wipe-clean this directory before
starting, see also the command <strong>--run_clean</strong>.</p>

<p><strong>--rulepath [path]</strong>, where [path] is the filepath to a text file listing
the fixed gene expression inputs and their corresponding regulatory problems.</p>

<p><strong>--psampath [path]</strong>, where [path] is the filepath to a text file describing
the position-specific affinity matrices (i.e., PSAMs) for all transcription factors
in one genome.  Every individual in the population will be initialized with these PSAMs.
Note: If you the option --poppath is used (see below), then --psampath is not required.</p>

<p><strong>--urspath [path]</strong>, where [path] is the filepath to a text file containing
the upstream regulatory sequences (i.e. URSs) for all genes in one genome.  Every individual in the
population will be initizlied with these URSs. Note: If you the option --poppath is used (see below), then --urspath is not required.</p>

<p><strong>--poppath [path]</strong>, where [path] is the filepath to a saved population written
by SimGenome.  If this option is present, it will override values for --psampath and --urspath.  The
option --poppath is often combined with the option --startgen N, where N is not zero, in order
to continue a previously saved simulation at a later generation. Note: the option --poppath is only required
if the options --psampath and --urspath are NOT specified.</p>

<h2>Optional Parameters:</h2>

<p><strong>--verbosity N</strong>, where N can range from 0 to 100.
Higher values of N produce more output on the screen, while lower values produce
less output.  Values of 30 and higher
will produce very verbose output files that will reduce the performance
of SimGenome, although this output can be useful for comprehensively analyzing
your simulated conditions.  The default value is N = 3.</p>

<p><strong>--runclean</strong> will explicitly remove any existing files in the output
directory before the simulation begins.  This is useful if you are re-running a previous
analysis and want to prevent stale output from polluting your results.</p>

<p><strong>--cooppath [path]</strong>, where [path] is the filepath to a text file
containing cofactor interaction definitions.  If this file is not specified, then all
co-factor interactions will be initialized to 1.0, indicating no specific interaction.  
See the examples for more information about formatting this file.</p>

<p><strong>--startgen N</strong> will start the simulation at generation N. The default value is 0.</p>

<p><strong>--maxgen N</strong> will stop the simulation when it reaches generation N.
The default value is 1000.</p>

<p><strong>--popsize N</strong> will create a population of size N.  Each individual
will be initialized as a clone of the URSs, PSAMs, and co-factor interactions described
in --urspath, --psampath, and --cooppath (if specified).  This parameter is
ignored if you also specify --poppath.</p>

<p><strong>--maxgd N</strong>, where N is the maximum distance, in nucleotide sites, over
which two co-factors can interact.  Co-factor interactions are strongest when the
co-factors are nearby, and the interaction degrades over distance.  The default value is 3 sites.</p>

<p><strong>--niid N</strong> will set the number of independent and identically
distributed (I.I.D.) samples to draw from the distribution of binding occupancies.  Higher
values will increase the accuracy of SimGenome, but will also take longer to compute.  The default
value is 5000.</p>

<p><strong>--pe_scalar X</strong>, where X is the expression scalar.
Higher values make it easer for genes to be expressed with fewer sites bound in their URS,
while lower values mean that expression requires more sites bound in the URS.
The default value is 0.005. The relationship between pe_scalar and expression levels is as follows.
Expression for some gene at time t+1 equals expression at time t plus a modifier pe. The modifier is computed
using the following expression:<br>
pe = R * [ (1 / (1+exp(-1*ss->pe_scalar*(sum_act-sum_rep) ) ) ) - 0.5]<br>
...where R is the growth rate, or decay rate

</p>

<p><strong>--f_scalar X</strong>, where X is the fitness scalar used to translate an individual's expression
error into a fitness score, according to the following equation: [fitness = e^(f_scalar x error)]. The default value
for the scalar is -0.0005. Values approaching zero will create a broad fitness function, while values much lower than zero create
a sharper fitness peak. The relationship between f_scalar, expression error, and fitness are shown in the plot below. It is recommended
that you tune this parameter according to the particular regulatory problems in your simulation, such that the fitness function can accomodate
the maximum possible expression error.</p>
<div align="center">
<img src="graphics/f_scalar.png" width="50%">
</div>

<p><strong>--log_occupancy</strong> enables binding occupancies to be logged. This log consumes a lot of disk space,
but is useful for determining the logical circuit of an individual. Logs written by this feature will appear in an output
 folder titled 'OCCUPANCY'.
</p>

<p><strong>--log_k</strong> enables activation and repression energies to be logged. This log provides a more compact and summarized version
of the data in the occupancy log (see above, for '--log_occupancy'). However, unlike the occupancy log, the K log doesn't include
information for every sequence site. Rather, the K log expresses energies over the entire regulatory sequence. This log is
especially useful for inferring the logical structure of the circuit. 
</p>

<h3>Basic Mutations:</h3>

<p><strong>--nomu</strong> will disable all mutations in the genetic algorithm. By default, mutations are enabled.</p>

<p><strong>--urs_mu X</strong> sets the mean mutation rate for URSs to X, where X is
nucleotide substitutions per sequence site. Default = 0.01.</p>

<p><strong>--psam_mu X</strong> sets the mean mutation rate for PSAMs to X, where X is the
probability of mutating a single site in a PSAM. Default = 0.05.</p>

<p><strong>--mu_stdev X</strong> sets the standard deviation of the mutation rates for URSs and PSAMs.
Default = 0.01.
</p>

<h3>Insertion/Deletion:</h3>

<p><strong>--urslenmu X</strong>, where X is the probability of inserting or deleting sites
into each URS. Default = 0.05.</p>

<p><strong>--urslensd X</strong>, where X is the standard deviation of the indel mutation rate provided
by 'urslenmu'. Default = 0.01.</p>

<p><strong>--urslensizemu X</strong>, where X is the mean URS indel size. Default = 3 sites.</p>

<p><strong>--urslensizesd X</strong>, where X is the standard deviation of the mean URS indel size. Default = 1 site.</p>

<p><strong>--psamlenmu X</strong>, where X is the probability of mutating the length of each PSAM.
The mutation can be an insertion or a deletion of a PSAM site. Default = 0.05.</p>

<p><strong>--psamlensd X</strong>, where X is the standard deviation of the PSAM indel mutation rate. Default = 0.01.</p>

<p><strong>--psamlensizemu X</strong>, were X is the mean PSAM indel size. Default = 2 sites.</p>

<p><strong>--psamlensizesd X</strong>, where X is the standard deviation of the mean PSAM indel size. Default = 1 site.</p>

<h3>Co-Factor Interaction Mutations:</h3>

<p><strong>--coopmu X</strong>, where X is the probability of mutating co-factor interactions. Default = 0.0.</p>

<p><strong>--coopmusd X</strong>, where X is the standard deviation of the probability of mutating co-factor interactions. Default = 0.0</p>

<p><strong>--coopmusize X</strong>, where X is the mean multiplier by which each mutated 
co-factor mutation will increase or decrease. Default = 1.1.</p>

<hr>

</BODY>
</HTML>